### PR TITLE
Improve code generation for the JIT

### PR DESCRIPTION
**Pull request by bjorng**
This pull request introduces many small improvements to the code generation for the JIT.

### CODE REVIEW

File: erts/emulator/beam/jit/arm/instr_common.cpp [hunk 1]
```diff
@@ -1296,17 +1296,9 @@ void BeamModuleAssembler::emit_i_is_tagged_tuple(const ArgLabel &Fail,
...
-               imm(NZCV::kNone),
-               imm(arm::CondCode::kEQ));
-    } else {
-        cmp_arg(TMP1, Arity);
-        a.b_ne(resolve_beam_label(Fail, disp1MB));
-        cmp_arg(TMP2, Tag);
-    }
+    cmp_arg(TMP2, Tag);
+    mov_imm(TMP3, Arity.get());
+    a.ccmp(TMP1, TMP3, imm(NZCV::kNone), imm(arm::CondCode::kEQ));
```
Comment (jhogberg) [inline erts/emulator/beam/jit/arm/instr_common.cpp:1316]: Isn't this sequence one instruction longer when the arity is <32? Perhaps we should make a `ccmp` helper that uses immediates when possible?

File: erts/emulator/beam/jit/arm/instr_common.cpp [hunk 2]
```diff
@@ -1463,7 +1463,8 @@ void BeamModuleAssembler::emit_is_eq_exact(const ArgLabel &Fail,
     a.cmp(x.reg, y.reg);
     a.b_eq(next);
 
-    if (exact_type<BeamTypeId::Integer>(X) && exact_type<BeamTypeId::Integer>(X)) {
+    if (exact_type<BeamTypeId::Integer>(X) &&
+        exact_type<BeamTypeId::Integer>(X)) {
```
Comment (jhogberg) [inline erts/emulator/beam/jit/arm/instr_common.cpp]: Shouldn't this be `X` and `Y`, not `X` and `X`?

File: erts/emulator/beam/jit/x86/instr_common.cpp [hunk 3]
```diff
@@ -784,10 +785,30 @@ void BeamModuleAssembler::emit_put_tuple2(const ArgRegister &Dst,
...
             }
-            case ArgVal::none:
-                mov_arg(dst_ptr, args[i]);
+            case ArgVal::none: {
+                unsigned j;
+                if (value == args[i]) {
+                    a.mov(dst_ptr, RET);
+                    break;
+                }
+                for (j = i + 1; j < size && args[i] == args[j]; j++) {
```
Comment (jhogberg) [inline erts/emulator/beam/jit/x86/instr_common.cpp:794]: `std::adjacent_find` (inverted)?

File: erts/emulator/beam/jit/arm/beam_asm.hpp [hunk 4]
```diff
@@ -1532,6 +1532,19 @@ class BeamModuleAssembler : public BeamAssembler,
...
+    void flush_vars(const Variable<arm::Gp> &to1,
+                    const Variable<arm::Gp> &to2,
+                    const Variable<arm::Gp> &to3) {
+        if (!to1.mem.hasBaseReg()) {
+            flush_vars(to2, to3);
+        } else if (!to2.mem.hasBaseReg()) {
+            flush_vars(to1, to3);
+        } else {
+            flush_vars(to1, to2);
+            flush_var(to3);
```
Comment (jhogberg) [inline erts/emulator/beam/jit/arm/beam_asm.hpp:1589]: What if `to1` (or `to2`) and `to3` are adjacent, but `to1` and `to2` are not? Should we check that?

File: erts/emulator/beam/jit/arm/beam_asm.hpp [hunk 5]
```diff
@@ -1329,6 +1348,10 @@ class BeamModuleAssembler : public BeamAssembler,
      * every `STUB_CHECK_INTERVAL` bytes for veneers and constants to work. */
     void check_pending_stubs();
 
+    /* Unconditionally emits all pending labels. Must only be called when
+     * the current code position is uncreachble. */
```
Comment (jhogberg) [inline erts/emulator/beam/jit/arm/beam_asm.hpp:1353]: Nitpick: unreachable.

File: erts/emulator/beam/jit/arm/instr_select.cpp [hunk 6]
```diff
@@ -478,52 +507,65 @@ void BeamModuleAssembler::emit_i_jump_on_val(const ArgSource &Src,
...
+            cmp(reg, y);
+            a.ccmp(reg, imm(x), imm(NZCV::kEqual), imm(arm::CondCode::kNE));
+        } else if (-y < 32) {
+            cmp(reg, x);
+            a.ccmn(reg, imm(-y), imm(NZCV::kEqual), imm(arm::CondCode::kNE));
+        } else {
+            cmp(reg, x);
+            a.mov(tmp, y);
+            a.ccmp(reg, tmp, imm(NZCV::kEqual), imm(arm::CondCode::kNE));
+        }
```
Comment (jhogberg) [inline erts/emulator/beam/jit/arm/instr_select.cpp:568]: This looks like a dead ringer for the `ccmp` helper I mentioned in the `is_tagged_tuple` commit.

File: erts/emulator/beam/jit/arm/instr_common.cpp (hunk 1 above)
Comment (bjorng) [reply erts/emulator/beam/jit/arm/instr_common.cpp:1316]: It is the tagged header word. It is never less than 32.

File: erts/emulator/beam/jit/arm/beam_asm.hpp (hunk 4 above)
Comment (bjorng) [reply erts/emulator/beam/jit/arm/beam_asm.hpp:1589]: I will rewrite this to check for adjacent memory addresses.

File: erts/emulator/beam/jit/arm/instr_select.cpp (hunk 6 above)
Comment (bjorng) [reply erts/emulator/beam/jit/arm/instr_select.cpp:568]: Looking at `ccmp` instructions, this one seems to be the only one that would need that helper.

File: erts/emulator/beam/jit/arm/instr_common.cpp (hunk 2 above)
Comment (bjorng) [reply erts/emulator/beam/jit/arm/instr_common.cpp]: Yes.

File: erts/emulator/beam/jit/arm/instr_bif.cpp [hunk 7]
```diff
@@ -568,15 +571,17 @@ void BeamGlobalAssembler::emit_call_light_bif_shared() {
...
     Label entry = a.newLabel();
+    BeamFile_ImportEntry *e = &beam->imports.entries[Exp.get()];
 
     a.bind(entry);
 
     mov_arg(ARG4, Exp);
     mov_arg(ARG8, Bif);
     a.adr(ARG3, entry);
 
+    ERTS_ASSERT(e->arity <= MAX_BIF_ARITY);
```
[the remark this answers is not in the document]
Comment (bjorng) [reply erts/emulator/beam/jit/arm/instr_bif.cpp]: Done.

**Pull request merged**